src/services/http/controllers/ZkChat.ts

```diff
@@ -0,0 +1,175 @@
+    res.send(makeResponse(users));
+  };
+
+  postMessage = async (req: Request, res: Response) => {
+    const { messageId, type, timestamp, sender, receiver, ciphertext, rln, semaphore } = req.body;
+    const signature = req.header('X-SIGNED-ADDRESS');
+    const userDB = await this.call('db', 'getUsers');
+
+    if (!sender.address && !sender.hash && !sender.ecdh) throw new Error('invalid sender');
+    if (!receiver.address && !receiver.ecdh) throw new Error('invalid receiver');
+
+    if (rln) {
+      if (!sender.hash) throw new Error('invalid request object');
+
+      const isEpochCurrent = await this.call('zkchat', 'isEpochCurrent', rln.epoch);
+      const verified = await this.call('zkchat', 'verifyRLNProof', rln);
+      const root = '0x' + BigInt(rln.publicSignals.merkleRoot).toString(16);
+      const group = await this.call('merkle', 'getGroupByRoot', root);
+
+      if (!isEpochCurrent) throw new Error('outdated message');
+      if (!verified) throw new Error('invalid rln proof');
+      if (!group) throw new Error('invalid merkle root');
+
+      rln.group_id = group;
+
+      const share = {
+        nullifier: rln.publicSignals.internalNullifier,
+        epoch: rln.publicSignals.epoch,
+        y_share: rln.publicSignals.yShare,
+        x_share: rln.x_share,
+      };
+
+      const { isSpam, isDuplicate } = await this.call('zkchat', 'checkShare', share);
+
+      if (isDuplicate) {
+        throw new Error('duplicate message');
+      }
+
+      if (isSpam) {
+        res.status(429).send('too many requests');
+        return;
+      }
+
+      await this.call('zkchat', 'insertShare', share);
+      await this.merkleRoot?.addRoot(root, group);
+    } else if (semaphore) {
+      const verified = await this.call('zkchat', 'verifySemaphoreProof', semaphore);
+      const root = '0x' + BigInt(semaphore.publicSignals.merkleRoot).toString(16);
+      const group = await this.call('merkle', 'getGroupByRoot', root);
+      if (!verified) throw new Error('invalid proof');
+      if (!group) throw new Error('invalid merkle root');
+      await this.merkleRoot?.addRoot(root, group);
+    } else if (signature) {
+      const [sig, address] = signature.split('.');
+      const user = await userDB.findOneByName(address);
+
+      if (user?.pubkey) {
+        if (!verifySignatureP256(user.pubkey, address, sig)) {
+          res.status(403).send(makeResponse('unauthorized', true));
+          return;
+        }
+      }
+    } else {
+      res.status(403).send(makeResponse('unauthorized', true));
+      return;
+    }
+
+    const data = await this.call('zkchat', 'addChatMessage', {
+      messageId,
+      type,
+      timestamp: new Date(timestamp),
+      sender,
+      receiver,
+      ciphertext,
+      rln,
+    });
+
+    // await
+    publishTopic(`ecdh:${data.sender_pubkey}`, {
+      type: SSEType.NEW_CHAT_MESSAGE,
+      message: data,
+    });
+    // await?
+    publishTopic(`ecdh:${data.receiver_pubkey}`, {
```

Comment:
i didn't do `await` since failing to publish a topic via SSE is more like a side effect, and the request itself should still return success